node/beacon/src/authority.rs

```diff
@@ -81,51 +90,106 @@ impl Authority {
-    pub fn new(authority_config: AuthorityConfig, blockchain: &BeaconBlockChain) -> Self {
+    pub fn new(
+        authority_config: AuthorityConfig,
+        blockchain: &BeaconBlockChain,
+        storage: Arc<RwLock<BeaconChainStorage>>
+    ) -> Self {
         // TODO: cache authorities in the Storage, to not need to process the whole chain.
         let mut result = Self {
             authority_config,
             proposals: HashMap::new(),
             participation: HashMap::new(),
             processed_blocks: HashMap::new(),
+            current_epoch: 0,
             thresholds: HashMap::new(),
             accepted_authorities: HashMap::new(),
+            storage,
         };
-        // Initial authorities operate for the first two epochs.
-        let (accepted_authorities, threshold) = result.compute_threshold_accepted(
-            &CryptoHash::default(),
-            result.authority_config.initial_proposals.to_vec(),
-            vec![],
-        );
-        let mut slot = 0;
-        for epoch in 0..=1 {
-            result.thresholds.insert(epoch, threshold);
-            for slot_auth in &accepted_authorities {
-                slot += 1;
-                result.accepted_authorities.insert(slot, slot_auth.to_vec());
+        result.read_from_storage();
+        if result.proposals.is_empty() {
+            // Initial authorities operate for the first two epochs.
+            let (accepted_authorities, threshold) = result.compute_threshold_accepted(
+                &CryptoHash::default(),
+                result.authority_config.initial_proposals.to_vec(),
+                vec![],
+            );
+            let mut slot = 0;
+            for epoch in 0..=1 {
+                result.thresholds.insert(epoch, threshold);
+                for slot_auth in &accepted_authorities {
+                    slot += 1;
+                    result.accepted_authorities.insert(slot, slot_auth.to_vec());
+                }
             }
-        }
-        // Catch up with the blockchain. Note, the last block is allowed to progress while we
-        // are iterating.
-        // TODO: Take care of the fork being changed while we are iterating.
-        let mut index = 1;
-        let mut last_progress = 101;
-        while index <= blockchain.best_block().header().body.index {
-            if log_enabled!(target: "client", Debug) {
-                let best_block_index = blockchain.best_block().header().body.index;
-                let progress = index * 100 / best_block_index;
-                if progress != last_progress {
-                    debug!(target: "client", "Processing blocks {} out of {}", index, best_block_index);
-                    last_progress = progress;
+            // Catch up with the blockchain. Note, the last block is allowed to progress while we
+            // are iterating.
+            // TODO: Take care of the fork being changed while we are iterating.
```

Comment:
we can have forks in two cases:
  - less then 2/3 sigs (e.g. block is not finalized, we should not process it here)
 - malicious 2/3 sigs on a block, and we have network split. Here we can actually can have switching of the preferred fork. But should think through this on across client in general.

Reply:
I think it is out of the scope of this PR